From: Alexander Mikhaylenko <alexander.mikhaylenko@puri.sm>
Date: Tue, 12 Jul 2022 02:32:37 +0400
Subject: Add PurismSettings

Note: Most of the included code of connecting to the portal was taken
from libadwaita.

Co-authored-by: Mohammed Sadiq <sadiq@sadiqpk.org>
Co-authored-by: Julian Sparber <julian@sparber.net>
---
 gtk/meson.build                                  |   2 +
 gtk/org.gtk.gtk4.Settings.Purism.gschema.xml     |  18 ++
 gtk/purism-settings.c                            | 311 +++++++++++++++++++++++
 gtk/{gtkheaderbarprivate.h => purism-settings.h} |  18 +-
 4 files changed, 344 insertions(+), 5 deletions(-)
 create mode 100644 gtk/org.gtk.gtk4.Settings.Purism.gschema.xml
 create mode 100644 gtk/purism-settings.c
 copy gtk/{gtkheaderbarprivate.h => purism-settings.h} (63%)

diff --git a/gtk/meson.build b/gtk/meson.build
index 03b0b05..f03715f 100644
--- a/gtk/meson.build
+++ b/gtk/meson.build
@@ -151,6 +151,7 @@ gtk_private_sources = files([
   'gtktextviewchild.c',
   'timsort/gtktimsort.c',
   'gtktrashmonitor.c',
+  'purism-settings.c',
 ])
 
 # List of files that contain public API, and should be introspected
@@ -1092,6 +1093,7 @@ gtk_settings_schemas = [
   'org.gtk.gtk4.Settings.EmojiChooser.gschema.xml',
   'org.gtk.gtk4.Settings.Debug.gschema.xml',
   'org.gtk.gtk4.Inspector.gschema.xml',
+  'org.gtk.gtk4.Settings.Purism.gschema.xml',
 ]
 
 install_data(gtk_settings_schemas, install_dir: gtk_schemasdir, install_tag: 'runtime')
diff --git a/gtk/org.gtk.gtk4.Settings.Purism.gschema.xml b/gtk/org.gtk.gtk4.Settings.Purism.gschema.xml
new file mode 100644
index 0000000..1e71b4f
--- /dev/null
+++ b/gtk/org.gtk.gtk4.Settings.Purism.gschema.xml
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<schemalist>
+
+  <schema id='org.gtk.gtk4.Settings.Purism' path='/org/gtk/gtk4/settings/purism/'>
+    <key name='is-phone' type='b'>
+      <default>false</default>
+      <summary>Enable hacks for small screens</summary>
+      <description>
+        If this setting is true, GTK enables various tweaks to let it
+        work on small screens.
+
+        This is a temporary workaround until automatic adaptive UIs
+        are implemented based on the workarea size.
+      </description>
+    </key>
+  </schema>
+
+</schemalist>
diff --git a/gtk/purism-settings.c b/gtk/purism-settings.c
new file mode 100644
index 0000000..586bea1
--- /dev/null
+++ b/gtk/purism-settings.c
@@ -0,0 +1,311 @@
+/*
+ * Copyright (c) 2022 Purism SPC
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "config.h"
+
+#include "purism-settings.h"
+
+#include "gtkprivate.h"
+
+#define PORTAL_BUS_NAME "org.freedesktop.portal.Desktop"
+#define PORTAL_OBJECT_PATH "/org/freedesktop/portal/desktop"
+#define PORTAL_SETTINGS_INTERFACE "org.freedesktop.portal.Settings"
+
+#define PORTAL_ERROR_NOT_FOUND "org.freedesktop.portal.Error.NotFound"
+
+struct _PurismSettings {
+  GObject parent_instance;
+
+  GSettings *settings;
+  GDBusProxy *portal;
+
+  gboolean is_phone;
+};
+
+enum {
+  PROP_0,
+  PROP_IS_PHONE,
+  LAST_PROP
+};
+
+static GParamSpec *props[LAST_PROP] = { NULL, };
+
+G_DEFINE_TYPE (PurismSettings, purism_settings, G_TYPE_OBJECT)
+
+static void
+set_is_phone (PurismSettings *self,
+              gboolean        is_phone)
+{
+  is_phone = !!is_phone;
+
+  if (is_phone == self->is_phone)
+    return;
+
+  self->is_phone = is_phone;
+
+  g_object_notify_by_pspec (G_OBJECT (self), props[PROP_IS_PHONE]);
+}
+
+static void
+notify_settings_cb (PurismSettings *self)
+{
+  g_assert (self->settings);
+
+  set_is_phone (self, g_settings_get_boolean (self->settings, "is-phone"));
+}
+
+static gboolean
+init_gsettings (PurismSettings *self)
+{
+  GSettingsSchema *schema;
+
+  /* Checking gsettings is not useful in flatpak */
+  if (gdk_running_in_sandbox ())
+    return FALSE;
+
+  schema = g_settings_schema_source_lookup (g_settings_schema_source_get_default (),
+                                            "org.gtk.gtk4.Settings.Purism", TRUE);
+
+  if (!schema)
+    return FALSE;
+
+  self->settings = g_settings_new_full (schema, NULL, NULL);
+
+  if (self->settings)
+    {
+      notify_settings_cb (self);
+
+      g_signal_connect_object (self->settings,
+                              "changed::is-phone",
+                              G_CALLBACK (notify_settings_cb),
+                              self,
+                              G_CONNECT_SWAPPED);
+    }
+
+  g_settings_schema_unref (schema);
+
+  return TRUE;
+}
+
+static gboolean
+read_portal_setting (PurismSettings  *self,
+                     const char      *schema,
+                     const char      *name,
+                     const char      *type,
+                     GVariant       **out)
+{
+  GError *error = NULL;
+  GVariant *ret;
+  GVariant *child, *child2;
+  GVariantType *out_type;
+  gboolean result = FALSE;
+
+  g_assert (self->portal);
+
+  ret = g_dbus_proxy_call_sync (self->portal,
+                                "Read",
+                                g_variant_new ("(ss)", schema, name),
+                                G_DBUS_CALL_FLAGS_NONE,
+                                G_MAXINT,
+                                NULL,
+                                &error);
+  if (error)
+    {
+      if (error->domain == G_DBUS_ERROR &&
+          error->code == G_DBUS_ERROR_SERVICE_UNKNOWN)
+        g_debug ("Portal not found: %s", error->message);
+      else if (error->domain == G_DBUS_ERROR &&
+                 error->code == G_DBUS_ERROR_UNKNOWN_METHOD)
+        g_debug ("Portal doesn't provide settings: %s", error->message);
+      else if (g_dbus_error_is_remote_error (error))
+        {
+          char *remote_error = g_dbus_error_get_remote_error (error);
+
+          if (!g_strcmp0 (remote_error, PORTAL_ERROR_NOT_FOUND))
+            g_debug ("Setting %s.%s of type %s not found", schema, name, type);
+
+          g_free (remote_error);
+        }
+      else
+        g_critical ("Couldn't read the %s setting: %s", name, error->message);
+
+      g_clear_error (&error);
+
+      return FALSE;
+    }
+
+  g_variant_get (ret, "(v)", &child);
+  g_variant_get (child, "v", &child2);
+
+  out_type = g_variant_type_new (type);
+  if (g_variant_type_equal (g_variant_get_type (child2), out_type))
+    {
+      *out = child2;
+
+      result = TRUE;
+    }
+  else
+    {
+      g_critical ("Invalid type for %s.%s: expected %s, got %s",
+                  schema, name, type, g_variant_get_type_string (child2));
+
+      g_variant_unref (child2);
+    }
+
+  g_variant_type_free (out_type);
+  g_variant_unref (child);
+  g_variant_unref (ret);
+  g_clear_error (&error);
+
+  return result;
+}
+
+static void
+portal_changed_cb (GDBusProxy     *proxy,
+                   const char     *sender_name,
+                   const char     *signal_name,
+                   GVariant       *parameters,
+                   PurismSettings *self)
+{
+  GVariant *is_phone = NULL;
+  const char *namespace;
+  const char *name;
+
+  if (g_strcmp0 (signal_name, "SettingChanged"))
+    return;
+
+  g_variant_get (parameters, "(&s&sv)", &namespace, &name, &is_phone);
+
+  if (!g_strcmp0 (namespace, "org.gtk.gtk4.Settings.Purism") &&
+      !g_strcmp0 (name, "is-phone"))
+    set_is_phone (self, g_variant_get_boolean (is_phone));
+
+  g_variant_unref (is_phone);
+}
+
+static gboolean
+init_portal (PurismSettings *self)
+{
+  GError *error = NULL;
+  GVariant *is_phone;
+
+  self->portal = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
+                                                G_DBUS_PROXY_FLAGS_NONE,
+                                                NULL,
+                                                PORTAL_BUS_NAME,
+                                                PORTAL_OBJECT_PATH,
+                                                PORTAL_SETTINGS_INTERFACE,
+                                                NULL,
+                                                &error);
+  if (error)
+    {
+      g_debug ("Settings portal not found: %s", error->message);
+
+      g_error_free (error);
+
+      return FALSE;
+    }
+
+  if (!read_portal_setting (self, "org.gtk.gtk4.Settings.Purism",
+                           "is-phone", "b", &is_phone))
+    return FALSE;
+
+  set_is_phone (self, g_variant_get_boolean (is_phone));
+  g_variant_unref (is_phone);
+
+  g_signal_connect (self->portal, "g-signal",
+                    G_CALLBACK (portal_changed_cb), self);
+
+  return TRUE;
+}
+
+static void
+purism_settings_dispose (GObject *object)
+{
+  PurismSettings *self = PURISM_SETTINGS (object);
+
+  g_clear_object (&self->portal);
+  g_clear_object (&self->settings);
+
+  G_OBJECT_CLASS (purism_settings_parent_class)->dispose (object);
+}
+
+static void
+purism_settings_get_property (GObject    *object,
+                              guint       prop_id,
+                              GValue     *value,
+                              GParamSpec *pspec)
+{
+  PurismSettings *self = PURISM_SETTINGS (object);
+
+  switch (prop_id)
+    {
+    case PROP_IS_PHONE:
+      g_value_set_boolean (value, purism_settings_get_is_phone (self));
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+    }
+}
+
+static void
+purism_settings_class_init (PurismSettingsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->dispose = purism_settings_dispose;
+  object_class->get_property = purism_settings_get_property;
+
+  props[PROP_IS_PHONE] =
+      g_param_spec_boolean ("is-phone", NULL, NULL,
+                            FALSE,
+                            GTK_PARAM_READABLE);
+
+  g_object_class_install_properties (object_class, LAST_PROP, props);
+}
+
+static void
+purism_settings_init (PurismSettings *self)
+{
+  if (init_gsettings (self))
+    return;
+
+  if (gdk_display_should_use_portal (gdk_display_get_default (), PORTAL_SETTINGS_INTERFACE, 0))
+    init_portal (self);
+}
+
+PurismSettings *
+purism_settings_get_default (void)
+{
+  static PurismSettings *settings = NULL;
+
+  if (!settings)
+    settings = g_object_new (PURISM_TYPE_SETTINGS, NULL);
+
+  return settings;
+}
+
+gboolean
+purism_settings_get_is_phone (PurismSettings *self)
+{
+  g_return_val_if_fail (PURISM_IS_SETTINGS (self), FALSE);
+
+  return self->is_phone;
+}
diff --git a/gtk/gtkheaderbarprivate.h b/gtk/purism-settings.h
similarity index 63%
copy from gtk/gtkheaderbarprivate.h
copy to gtk/purism-settings.h
index d7c924e..a7a796f 100644
--- a/gtk/gtkheaderbarprivate.h
+++ b/gtk/purism-settings.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2013 Red Hat, Inc.
+ * Copyright (c) 2022 Purism SPC
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU Lesser General Public License as published by
@@ -14,16 +14,24 @@
  * You should have received a copy of the GNU Lesser General Public License
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
- *
  */
 
 #pragma once
 
-#include "gtkheaderbar.h"
+#if !defined (__GTK_H_INSIDE__) && !defined (GTK_COMPILATION)
+#error "Only <gtk/gtk.h> can be included directly."
+#endif
+
+#include <glib-object.h>
 
 G_BEGIN_DECLS
 
-void         _gtk_header_bar_track_default_decoration (GtkHeaderBar *bar);
+#define PURISM_TYPE_SETTINGS (purism_settings_get_type ())
 
-G_END_DECLS
+G_DECLARE_FINAL_TYPE (PurismSettings, purism_settings, PURISM, SETTINGS, GObject)
 
+PurismSettings *purism_settings_get_default  (void);
+
+gboolean        purism_settings_get_is_phone (PurismSettings *self);
+
+G_END_DECLS
